File: arraysAndHashing/longestCommonPrefix/longestCommonPrefix.js
// Write a function to find the longest common prefix string amongst an
// array of strings.

// If there is no common prefix, return an empty string "".

// source: https://leetcode.com/problems/longest-common-prefix

var longestCommonPrefix = function(strs) {
    if (strs.length === 1) return strs[0];
    let res = strs[0]; // "flower"

    for (let word of strs) {
        for (let i = res.length - 1; i >= 0; i--) {
            if (res[i] !== word[i]) {
                res = res.slice(0, i);
            }
        }
    }

    return res;
};

File: arraysAndHashing/longestCommonPrefix/longest_common_prefix.py
class Solution(object):
    def longestCommonPrefix(self, strs):
        """
        :type strs: List[str]
        :rtype: str
        """
        if not strs:
            return ""

        prefix = strs[0]
        for string in strs[1:]:
            i = 0
            while i < len(prefix) and i < len(string) and prefix[i] == string[i]:
                i += 1
            prefix = prefix[:i]

            if not prefix:
                break

        return prefix

# s = sum of all characters in string
# time: O(s)
# space: O(1)
